### PR TITLE
implement health check in docker cluster, event based

### PR DESCRIPTION
**Pull request by runshenzhu**
Only the last commit is for health check.

Listening on an event queue to monitor container's status. Previously, Cluster doesn't have event method, `SubscribeToEvent` is introduced to implement `adapter.events`.

### CODE REVIEW

File: daemon/cluster/executor/container/adapter.go [hunk 1]
```diff
@@ -162,9 +164,47 @@ func (c *containerAdapter) inspect(ctx context.Context) (types.ContainerJSON, er
...
+
+	for _, event := range buffer {
+		eventsq <- event
+	}
+
+	go func() {
+		defer c.backend.UnsubscribeFromEvents(l)
+
+		select {
+		case <-ctx.Done():
```
Comment (dongluochen) [inline daemon/cluster/executor/container/adapter.go]: I'm not sure if this is needed. It's already handled in the `for` loop below.

File: daemon/cluster/executor/container/adapter.go [hunk 2]
```diff
@@ -162,9 +164,47 @@ func (c *containerAdapter) inspect(ctx context.Context) (types.ContainerJSON, er
...
+			select {
+			case ev := <-l:
+				jev, ok := ev.(events.Message)
+				if !ok {
+					log.G(ctx).Warnf("unexpected event message: %q", ev)
+					continue
+				}
+				select {
+				case eventsq <- jev:
+				case <-ctx.Done():
```
Comment (dongluochen) [inline daemon/cluster/executor/container/adapter.go:189]: Similar comment here.

File: daemon/cluster/executor/container/adapter.go (hunk 1 above)
Comment (runshenzhu) [inline daemon/cluster/executor/container/adapter.go]: I will remove it. Here I just implement it based on stevvooe's code.

File: daemon/cluster/executor/container/adapter.go (hunk 2 above)
Comment (runshenzhu) [inline daemon/cluster/executor/container/adapter.go:189]: I think `select` is necessary here. If ctx is done, `eventsq <- jev` may block and cause the leak of go routine.

**Comment by runshenzhu** [comment]
Can I send this PR to docker master? Or we should hold it for the further discussion?

**Comment by dongluochen** [comment]
Please send it. There might be more work to make it work end-to-end. But let's get feedbacks.

**Comment by runshenzhu** [comment]
@dongluochen Thanks for the review.

**Pull request closed without merging**